src/main/kotlin/tanvd/grazi/ide/ui/GraziRulesTree.kt

```diff
@@ -0,0 +1,279 @@
+            override fun nodeStateChanged(node: CheckedTreeNode) {
+                val meta = node.userObject
+                if (meta is RuleWithLang) {
+                    meta.state = node.isChecked
+                    if (meta.enabled == meta.state) {
+                        state.remove(meta.rule.id)
+                    } else {
+                        state[meta.rule.id] = meta
+                    }
+                }
+            }
+        })
+
+        val scrollPane = ScrollPaneFactory.createScrollPane(tree)
+        val filterPanel = JPanel(BorderLayout())
+        filterPanel.add(filter, BorderLayout.CENTER)
+        filterPanel.border = JBUI.Borders.emptyBottom(2)
+
+        val group = DefaultActionGroup()
+        val actionManager = CommonActionsManager.getInstance()
+        val treeExpander = DefaultTreeExpander(tree)
+        group.add(actionManager.createExpandAllAction(treeExpander, tree))
+        group.add(actionManager.createCollapseAllAction(treeExpander, tree))
+        filterPanel.add(ActionManager.getInstance().createActionToolbar("GraziRulesTree", group, true).component, BorderLayout.WEST)
+
+        panel.add(filterPanel, BorderLayout.NORTH)
+        panel.add(scrollPane, BorderLayout.CENTER)
+
+        filter.reset()
+    }
+
+    private fun getNodeText(node: CheckedTreeNode): String {
+        return when (val meta = node.userObject) {
+            is RuleWithLang -> meta.rule.description
+            is Category -> meta.name
+            is Lang -> meta.displayName
+            else -> "???"
+        }
+    }
+
+    private fun visit(node: CheckedTreeNode, visitor: (node: CheckedTreeNode) -> Unit) {
+        node.children().iterator().forEach {
+            visitor(it as CheckedTreeNode)
+        }
+    }
+
+    fun apply() {
+        state.values.forEach { rule ->
+            if (rule.state) {
+                LangTool[rule.lang].enableRule(rule.rule.id)
+                GraziConfig.state.userDisabledRules.remove(rule.rule.id)
+                GraziConfig.state.userEnabledRules.add(rule.rule.id)
+            } else {
+                LangTool[rule.lang].disableRule(rule.rule.id)
+                GraziConfig.state.userEnabledRules.remove(rule.rule.id)
+                GraziConfig.state.userDisabledRules.add(rule.rule.id)
+            }
+        }
+
+        state.clear()
+    }
+
+    private fun resetCheckMark(root: CheckedTreeNode): Boolean {
+        val meta = root.userObject
+        if (meta is RuleWithLang) {
+            root.isChecked = when(val rule = state[meta.rule.id]) {
+                is RuleWithLang -> rule.state
+                else -> !LangTool[meta.lang].disabledRules.contains(meta.rule.id)
+            }
+        } else {
+            root.isChecked = false
+            visit(root) { if (resetCheckMark(it)) root.isChecked = true }
+        }
+
+        return root.isChecked
+    }
+
+    fun filter(filterString: String?) {
+
+        val filteredRules = LangTool.allRulesWithLangs()
```

Comment:
Try to rewrite it using `filter` and `flatMap`. It should be possible. And it will look a lot easier. Use `asSequence()` to be sure, that it will be filtered only once

Reply:
Done